File: js/test/entity/label/LabelEntity.test.js

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

const Path = require('node:path')
const Fs = require('node:fs')

const { test, describe } = require('node:test')
const assert = require('node:assert')


const { ShortcutSDK, BaseFeature, stdutil, config } = require('../../..')

const {
  envOverride,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
} = require('../../utility')


describe('LabelEntity', async () => {

  test('instance', async () => {
    const testsdk = ShortcutSDK.test()
    const ent = testsdk.Label()
    assert(null != ent)
  })


  test('basic', async () => {

    const setup = basicSetup()
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select


    // CREATE
    const label_ref01_ent = client.Label()
    let label_ref01_data = setup.data.new.label['label_ref01']

    label_ref01_data = (await label_ref01_ent.create(label_ref01_data)).data()
    assert(null != label_ref01_data.id)


    // LIST
    const label_ref01_match = {}

    const label_ref01_list = (await label_ref01_ent.list(label_ref01_match)).map((e) => e.data())

    assert(!isempty(select(label_ref01_list, { id: label_ref01_data.id })))


    // UPDATE
    const label_ref01_data_up0 = {}
    label_ref01_data_up0.id = label_ref01_data.id

    const label_ref01_markdef_up0 = { name: 'app_url', value: 'Mark01-label_ref01_' + setup.now }
    label_ref01_data_up0 [label_ref01_markdef_up0.name] = label_ref01_markdef_up0.value

    const label_ref01_resdata_up0 = (await label_ref01_ent.update(label_ref01_data_up0)).data()
    assert(label_ref01_resdata_up0.id === label_ref01_data_up0.id)

    assert(label_ref01_resdata_up0[label_ref01_markdef_up0.name] === label_ref01_markdef_up0.value)


    // LOAD
    const label_ref01_match_dt0 = {}
    label_ref01_match_dt0.id = label_ref01_data.id
    const label_ref01_data_dt0 = (await label_ref01_ent.load(label_ref01_match_dt0)).data()
    assert(label_ref01_data_dt0.id === label_ref01_data.id)


    // REMOVE
    const label_ref01_match_rm0 = {}
    label_ref01_match_rm0.id = label_ref01_data.id
    await label_ref01_ent.remove(label_ref01_match_rm0)
  

    // LIST
    const label_ref01_match_rt0 = {}

    const label_ref01_list_rt0 = (await label_ref01_ent.list(label_ref01_match_rt0)).map((e) => e.data())

    assert(isempty(select(label_ref01_list_rt0, { id: label_ref01_data.id })))


  })
})



function basicSetup(extra) {
  // TODO: fix test def options
  const options = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname,
      '../../../../.sdk/test/entity/label/LabelTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = ShortcutSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['label01','label02','label03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'SHORTCUT_TEST_LABEL_ENTID': idmap,
    'SHORTCUT_TEST_LIVE': 'FALSE',
    'SHORTCUT_TEST_EXPLAIN': 'FALSE',
    'SHORTCUT_APIKEY': 'NONE',
  })

  idmap = env['SHORTCUT_TEST_LABEL_ENTID']

  if ('TRUE' === env.SHORTCUT_TEST_LIVE) {
    client = new ShortcutSDK(merge([
      {
        apikey: env.SHORTCUT_APIKEY,
      },
      extra
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.SHORTCUT_TEST_EXPLAIN,
    now: Date.now(),
  }

  return setup
}
  

File: go/test/label_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/shortcut-sdk/go"
	"github.com/voxgig-sdk/shortcut-sdk/go/core"

	vs "github.com/voxgig-sdk/shortcut-sdk/go/utility/struct"
)

func TestLabelEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Label(nil)
		if ent == nil {
			t.Fatal("expected non-nil LabelEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"label": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Label(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Label(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := labelBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "label." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set SHORTCUT_TEST_LABEL_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		labelRef01Ent := client.Label(nil)
		labelRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "label"}), "label_ref01"))

		labelRef01DataResult, err := labelRef01Ent.Create(labelRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		labelRef01Data = core.ToMapAny(entityData(labelRef01DataResult))
		if labelRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if labelRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		labelRef01Match := map[string]any{}

		labelRef01ListResult, err := labelRef01Ent.List(labelRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		labelRef01List, labelRef01ListOk := labelRef01ListResult.([]any)
		if !labelRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", labelRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(labelRef01List), map[string]any{"id": labelRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		labelRef01DataUp0Up := map[string]any{
			"id": labelRef01Data["id"],
		}

		labelRef01MarkdefUp0Name := "app_url"
		labelRef01MarkdefUp0Value := fmt.Sprintf("Mark01-label_ref01_%d", setup.now)
		labelRef01DataUp0Up[labelRef01MarkdefUp0Name] = labelRef01MarkdefUp0Value

		labelRef01ResdataUp0Result, err := labelRef01Ent.Update(labelRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		labelRef01ResdataUp0 := core.ToMapAny(entityData(labelRef01ResdataUp0Result))
		if labelRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if labelRef01ResdataUp0["id"] != labelRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if labelRef01ResdataUp0[labelRef01MarkdefUp0Name] != labelRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", labelRef01MarkdefUp0Name, labelRef01ResdataUp0[labelRef01MarkdefUp0Name])
		}

		// LOAD
		labelRef01MatchDt0 := map[string]any{
			"id": labelRef01Data["id"],
		}
		labelRef01DataDt0Loaded, err := labelRef01Ent.Load(labelRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		labelRef01DataDt0LoadResult := core.ToMapAny(entityData(labelRef01DataDt0Loaded))
		if labelRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if labelRef01DataDt0LoadResult["id"] != labelRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		labelRef01MatchRm0 := map[string]any{
			"id": labelRef01Data["id"],
		}
		_, err = labelRef01Ent.Remove(labelRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		labelRef01MatchRt0 := map[string]any{}

		labelRef01ListRt0Result, err := labelRef01Ent.List(labelRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		labelRef01ListRt0, labelRef01ListRt0Ok := labelRef01ListRt0Result.([]any)
		if !labelRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", labelRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(labelRef01ListRt0), map[string]any{"id": labelRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func labelBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "label", "LabelTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read label test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse label test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"label01", "label02", "label03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("SHORTCUT_TEST_LABEL_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"SHORTCUT_TEST_LABEL_ENTID": idmap,
		"SHORTCUT_TEST_LIVE":      "FALSE",
		"SHORTCUT_TEST_EXPLAIN":   "FALSE",
		"SHORTCUT_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["SHORTCUT_TEST_LABEL_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["SHORTCUT_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["SHORTCUT_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewShortcutSDK(core.ToMapAny(mergedOpts))
	}

	live := env["SHORTCUT_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["SHORTCUT_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
